core/priority/calculate.go

```diff
@@ -0,0 +1,158 @@
+// individual peers may not contain duplicate topics, individual topics
+// may not contain duplicate priorities or more than 1000 priorities.
+//nolint:gocognit // Mostly just map reduce.
+func calculateResults(msgs []*pbv1.PriorityMsg, minRequired int) (*pbv1.PriorityResult, error) {
+	// Prepare deterministic input messages:
+	//  - validate
+	//  - then sort by peer
+	//  - then shuffle by slot (for random tiebreaker)
+	var (
+		slot       int64
+		input      []*pbv1.PriorityMsg
+		dedupPeers = newDeduper[string]("peer") // Peers may not be duplicated
+	)
+	for _, msg := range msgs {
+		if slot == 0 {
+			slot = msg.Slot
+		} else if msg.Slot != slot {
+			return nil, errors.New("mismatching slots")
+		}
+		if err := dedupPeers(msg.PeerId); err != nil {
+			return nil, err
+		}
+		dedupTopics := newDeduper[string]("topic") // Peers may not provide duplicate topics.
+		for _, topic := range msg.Topics {
+			if err := dedupTopics(topic.Topic); err != nil {
+				return nil, err
+			} else if len(topic.Priorities) >= maxPriorities {
+				return nil, errors.New("max priority reached")
+			}
+
+			dedupPriority := newDeduper[string]("priority") // Topics may not include duplicates priority.
+			for _, priority := range topic.Priorities {
+				if err := dedupPriority(priority); err != nil {
+					return nil, err
+				}
+			}
+		}
+
+		input = append(input, msg) // Copy to not mutate input param.
+	}
+	sort.Slice(input, func(i, j int) bool {
+		return input[i].PeerId < input[j].PeerId
+	})
+	//nolint:gosec // Math rand used for deterministic behaviour.
+	rand.New(rand.NewSource(slot)).Shuffle(len(input), func(i, j int) {
+		input[i], input[j] = input[j], input[i]
+	})
+
+	// Group all priority sets by topic
+	prioritySetsByTopic := make(map[string][][]string)
+	for _, msg := range input {
+		for _, topic := range msg.Topics {
+			prioritySetsByTopic[topic.Topic] = append(prioritySetsByTopic[topic.Topic], topic.Priorities)
+		}
+	}
+
+	// Calculate cluster wide resulting priorities by topic.
+	var topicResults []*pbv1.PriorityTopic
+	for topic, prioritySet := range prioritySetsByTopic {
+		if len(prioritySet) < minRequired {
+			// Shortcut since min require cannot be met.
+			continue
+		}
+
+		// Calculate overall score for all priorities in the topic
+		// which effectively orders by count then by overall priority.
+		var (
+			scores        = make(map[string]int)
+			allPriorities []string
+		)
+		for _, priorities := range prioritySet {
+			for order, label := range priorities {
+				if _, ok := scores[label]; !ok {
+					allPriorities = append(allPriorities, label)
+				}
+				scores[label] += countWeight - order // Equivalent to ordering by count then by priority
+			}
+		}
+
+		// Order by score decreasing
+		sort.Slice(allPriorities, func(i, j int) bool {
+			return scores[allPriorities[i]] > scores[allPriorities[j]]
+		})
+
+		// Extract scores with min required count.
+		minScore := (minRequired - 1) * countWeight
```

Comment:
why "-1"?

Reply:
because we subtract priority, all peers will have a score less than `count * countWeight`